src/aggregate_with_randomness.cc

```diff
@@ -0,0 +1,247 @@
+    bool &validate,
+    const Napi::CallbackInfo &info) {
+    Napi::Env env = info.Env();
+    Napi::Value array_value = info[0];
+    if (!array_value.IsArray()) {
+        Napi::TypeError::New(
+            env, "Must pass an array to aggregateWithRandomness")
+            .ThrowAsJavaScriptException();
+        return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+    }
+    Napi::Array array = array_value.As<Napi::Array>();
+    sets_length = array.Length();
+    if (sets_length == 0) {
+        Napi::TypeError::New(
+            env, "Empty array passed to aggregateWithRandomness")
+            .ThrowAsJavaScriptException();
+        return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+    }
+    sets.resize(sets_length);
+
+    Napi::Value validate_value = info[1];
+    if (!validate_value.IsUndefined()) {
+        if (!validate_value.IsBoolean()) {
+            Napi::TypeError::New(
+                env, "Must pass a boolean for validateSerialized")
+                .ThrowAsJavaScriptException();
+            return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+        }
+        validate = validate_value.As<Napi::Boolean>().Value();
+    }
+
+    for (uint32_t i = 0; i < sets_length; i++) {
+        sets.push_back({nullptr, nullptr, 0});
+        Napi::Value set_value = array[i];
+        if (!set_value.IsObject()) {
+            Napi::TypeError::New(
+                env,
+                "Must pass an array of objects to "
+                "aggregateWithRandomness")
+                .ThrowAsJavaScriptException();
+            return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+        }
+        Napi::Object set = set_value.As<Napi::Object>();
+
+        try {
+            blst_ts::PublicKey *to_aggregate = blst_ts::PublicKey::Unwrap(
+                set.Get("publicKey").As<Napi::Object>());
+            sets[i].pk = to_aggregate->point.get();
+        } catch (...) {
+            Napi::Error::New(
+                env,
+                "BLST_ERROR: Invalid PublicKey at index "s + std::to_string(i))
+                .ThrowAsJavaScriptException();
+            return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+        }
+
+        try {
+            Napi::Value val = set.Get("signature");
+            Napi::Uint8Array typed_array = val.As<Napi::Uint8Array>();
+            sets[i].sig_data = typed_array.Data();
+            sets[i].sig_length = typed_array.ByteLength();
+        } catch (...) {
+            Napi::Error::New(
+                env,
+                "BLST_ERROR: Invalid Signature at index "s + std::to_string(i))
+                .ThrowAsJavaScriptException();
+            return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+        }
+    }
+
+    return blst_ts::BLST_TS_ERROR::SUCCESS;
+}
+
+blst_ts::BLST_TS_ERROR aggregate_with_randomness(
+    blst::P1 &aggregate_key,
+    blst::P2 &aggregate_sig,
+    std::string &error_msg,
+    BlstTsAddon *module,
+    const std::vector<SignatureAndPublicKeySet> &sets,
+    const size_t &sets_length,
```

Comment:
1. `size_t` is primitive and there's no point in passing primitives by const reference - copying them is as cheap as it gets and it avoids indirection.
2. Is `sets_length` even needed? Shouldn't it be the same as `sets.size()`?

Reply:
> 1. `size_t` is primitive and there's no point in passing primitives by const reference - copying them is as cheap as it gets and it avoids indirection.

Sweet.  Updated.

> 2. Is `sets_length` even needed? Shouldn't it be the same as `sets.size()`?

Nope, was a segfault because of the double allocation I mentioned above.  Updated.